Java/core/src/main/java/com/amazon/randomcutforest/util/ArrayPacking.java

```diff
@@ -0,0 +1,176 @@
+                output[3 + used++] = code;
+                len += pack;
+            }
+            checkArgument(used + 3 == output.length, "incorrect state");
+            return output;
+        }
+
+    }
+
+    public static short[] decompressToShort(int[] inputArray, boolean decomress) {
+        if (inputArray == null) {
+            throw new IllegalStateException(" array packing invoked on null arrays");
+        }
+        if (inputArray.length < 3 || !decomress) {
+            short[] output = new short[inputArray.length];
+            for (int i = 0; i < inputArray.length; i++) {
+                output[i] = (short) inputArray[i];
+            }
+            return output;
+        }
+        int min = inputArray[0];
+        int max = inputArray[1];
+        short[] output = new short[inputArray[2]];
+        if (min == max) {
+            Arrays.fill(output, (short) min);
+        } else {
+            int base = (max - min + 1);
+            int pack = maxPack(base);
+            int count = 0;
+            for (int i = 3; i < inputArray.length; i++) {
+                int code = inputArray[i];
+                for (int j = 0; j < pack && count < output.length; j++) {
+                    output[count++] = (short) (min + code % base);
+                    code = code / base;
+                }
+            }
+            ;
+        }
+        return output;
+    }
+
+    public static int[] compress(int[] inputArray, boolean compress) {
+        if (inputArray == null) {
+            throw new IllegalStateException(" array packing invoked on null arrays");
+        }
+        if (!compress || inputArray.length < 3) {
+            return Arrays.copyOf(inputArray, inputArray.length);
+        }
+
+        int min = inputArray[0];
+        int max = inputArray[0];
+        for (int i = 1; i < inputArray.length; i++) {
+            min = Math.min(min, inputArray[i]);
+            max = Math.max(max, inputArray[i]);
+        }
+        int base = max - min + 1;
+        if (base == 1) {
+            return new int[] { min, max, inputArray.length };
+        } else {
+            int pack = maxPack(base);
+            int[] output = new int[3 + (int) Math.ceil(1.0 * inputArray.length / pack)];
+            output[0] = min;
+            output[1] = max;
+            output[2] = inputArray.length;
+            int len = 0;
+            int used = 0;
+            while (len < inputArray.length) {
+                int code = 0;
+                int reach = Math.min(len + pack - 1, inputArray.length - 1);
+                for (int i = reach; i >= len; i--) {
+                    code = base * code + (inputArray[i] - min);
+                }
+                output[3 + used++] = code;
+                len += pack;
+            }
+            checkArgument(used + 3 == output.length, "incorrect state");
+            return output;
+        }
+
+    }
+
+    public static int[] decompressToInt(int[] inputArray, boolean decomress) {
```

Comment:
If we change to `pack` we should also change this to `unpack`. I would also consider `unpackInts` and `unpackShorts` versus `unpackToInt` and `unpackToShort`.

Reply:
Done